Generic ephemeral volume API

pohly opened the pull request:
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:

Ephemeral inline volumes are useful for a variety of use cases (see https://github.com/kubernetes/enhancements/blob/a080b55b7237dd765ac828e58b7bb30a99755552/keps/sig-storage/1698-generic-ephemeral-volumes/README.md#user-stories), but the current solution is both CSI specific and limited (custom API, no storage capacity tracking, no volume resizing/snapshotting, etc.). For normal storage drivers and volumes, a different approach is needed.

**Which issue(s) this PR fixes**:

Related-To: https://github.com/kubernetes/enhancements/issues/1698

**Special notes for your reviewer**:

As part of the API review, we need to clarify:
- Is it enough that pod security policy can prevent users from using the new volume typo (https://github.com/kubernetes/enhancements/tree/a080b55b7237dd765ac828e58b7bb30a99755552/keps/sig-storage/1698-generic-ephemeral-volumes#risks-and-mitigations)?
- Should `EphemeralVolumeSource` embed a `PersistentVolumeTemplate` (as in the KEP) or is a `PersistentVolumeClaimSpec` sufficient (as in this PR)? This affects how labels are handled for the created PVC. I changed my mind while working on the API change and the implementation...

**Does this PR introduce a user-facing change?**:

```release-note
Introduce API for generic ephemeral inline volumes.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

Will be added together with the actual implementation.

/cc @liggitt 
/sig storage

pohly committed:
GenericEphemeralVolume: feature gate, API, documentation

As explained in
https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1698-generic-ephemeral-volumes,
CSI inline volumes are not suitable for more "normal" kinds of storage
systems. For those a new approach is needed: "generic ephemeral inline
volumes".

pohly commented on pkg/apis/core/types.go:
```diff
@@ -1670,6 +1673,29 @@ type CSIVolumeSource struct {
+	// to the PVC after it has been created.
+	//
+	// Required, must not be nil.
+	VolumeClaim *PersistentVolumeClaimSpec
```
In the KEP, I proposed to embed a full `PersistentVolumeClaim` (= `ObjectMeta` + `PersistentVolumeClaimSpec` + redundant `PersistentVolumeClaimStatus`, as in `StatefulSet`) which then led to a new `PersistentVolumeClaimTemplate` (= `ObjectMeta` + `PersistentVolumeClaimSpec`). My thinking at that time was that it would be useful to grant users full explicit control over the labels that get added to the created PVC.

However, that flexibility also has downsides, like the redundant name in the `ObjectMeta`. I then realized that tools like "kustomize" would have to be updated to deal correctly with explicit labels: currently, kustomize's `commonLabels` feature only adds labels to the meta data of types it knows about and doesn't add them to an embedded `PersistentVolumeClaim`.

Therefore I'd like to propose that we switch to one of the proposals during the KEP review and just embed a `PersistentVolumeClaimSpec`, with labels being copied from the pod meta data (implicit approach).

I'm open whether that should be *all* labels or whether we only copy certain labels, like those with the `app.kubernetes.io/` prefix (https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels).

@saad-ali: what do you think?

pohly replied:
@jsafrane already said that he prefers `PersistentVolumeTemplate` (= `ObjectMeta` + `PersistentVolumeClaimSpec`). Let me try to get to the bottom of what that means for the API:
- name, generatename and namespace are redundant
- all of the generated fields (UID, ResourceVersion, Generation, CreationTimestamp, DeletionTimestamp, DeletionGracePeriodSeconds) are redundant
- ownerreference is superseded or at least modified: does it make sense to allow more than one owner? I asked on Slack what multiple owners mean, but never fully understood that.
- Finalizers and ManagedFields don't make sense.
- The only fields which make sense are labels and annotations.

If we compose `PersistentVolumeTemplate` of `ObjectMeta`, should the user be allowed to set all fields, including those that don't make sense? I think the `StatefulSet` allows it. That may be a historic accident; we could decide to tighten validation such that for `PersistentVolumeClaimTemplate`, *only* labels and annotations can be set.

@jsafrane: agreed?

jsafrane replied:
I checked how Pod templates are validated in Deployments / StatefulSets etc, they do just very generic validation:

https://github.com/kubernetes/kubernetes/blob/442a69c3bdf6fe8e525b05887e57d89db1e2f3a5/pkg/apis/core/validation/validation.go#L4409-L4420

I think we should do the same - copy validation from PVC where it makes sense and don't restrict UID / Generation / CreationTimestamp / etc. The controller should only overwrite things it needs to control (GenerateName, Name).

---

BTW, I think we discussed this and you convinced me that we should use the same approach as StatefulSet, i.e. whole PVC object (incl. status) as a template:

https://github.com/kubernetes/kubernetes/blob/171b486c3483d763a06aa67ead4ff6e93d23874e/pkg/apis/apps/types.go#L130

I looked at how it's validated, but [it is not validated at all](https://github.com/kubernetes/kubernetes/blob/905e7510c84676fe5c6428f1038996e099a30f68/pkg/apis/apps/validation/validation.go#L71) (ehm...)

I see you current proposal, ClaimTemplate with ObjectMeta + Spec better.

pohly replied:
Okay, so to recap:
- we add `PersistentVolumeClaimTemplate` = `ObjectMeta` + `PersistentVolumeClaimSpec`
- only labels and annotations are copied by the controller, the rest of `ObjectMeta` is ignored
- validation is reused where it exists, but no further restrictions are enforced regarding the unused fields

Works for me. I'll extend the PR accordingly, using separate commits, so if we want to go back, we can.

liggitt replied:
> I checked how Pod templates are validated in Deployments / StatefulSets etc, they do just very generic validation

statefulsets are under-validated (xref https://github.com/kubernetes/kubernetes/issues/41153, https://github.com/kubernetes/kubernetes/issues/84698, https://github.com/kubernetes/kubernetes/issues/81147, https://github.com/kubernetes/kubernetes/issues/60619, https://github.com/kubernetes/kubernetes/issues/51306), let's not repeat those mistakes. I would expect us to reuse existing validation to ensure impossible-to-succeed things are not created.

> only labels and annotations are copied by the controller, the rest of ObjectMeta is ignored

persisting and ignoring data is misleading. I would expect to only allow persisting data in fields that will be honored.

> we add PersistentVolumeClaimTemplate = ObjectMeta + PersistentVolumeClaimSpec

what does this do to the size of the podspec schema? importing a complex type definition for spawning new objects into what has been our leaf object is a pretty notable change.

pohly replied:
> persisting and ignoring data is misleading. I would expect to only allow persisting data in fields that will be honored

We can specify that all other fields must be empty and I can add validation for that. However, that then leads to the question whether ObjectMeta is the right type to use in the first place.

`PersistentVolumeClaimTemplate` could also be defined as:
```
type PersistentVolumeClaimTemplate struct {
    Labels map[string]string
    Annotations map[string]string
    VolumeSpec PersistentVolumeClaimSpec
}
```

> what does this do to the size of the podspec schema?

I have pushed the change into the branch for this PR, so we can check that. Where do I find the podspec schema?

pohly replied:
> I'll extend the PR accordingly, using separate commits, so if we want to go back, we can.

I ended up archiving the previous source and just squashed my changes so (currently) 
77ad5d8 has the API as originally proposed in the KEP. But the defaulting test fails, so I'll have to revise that a bit.

jsafrane replied:
> persisting and ignoring data is misleading. I would expect to only allow persisting data in fields that will be honored.

We don'd to that for PodTemplates - validation allows UID, Generation, Name and maybe other stuff that's thrown away to be set there.

liggitt replied:
> We don'd to that for PodTemplates - validation allows UID, Generation, Name and maybe other stuff that's thrown away to be set there.

Yes, and that causes confusion and bug reports.

pohly replied:
I agree with @liggitt, the existing behavior isn't a good example to follow and I'd be happy to do something better - we just need to agree *what* that is. Embedding just `PersistentVolumeClaimSpec` was my attempt to avoid that entire source of confusion, but then there's no place to specify labels and annotations.

@liggitt : do you have some preferred solution?

To recap the alternatives:
- embed `PersistentVolumeClaimSpec`, labels (potentially limited to apps.kubernetes.io/*) copied from pod
- embed `ObjectMeta` + `PersistentVolumeClaimSpec`, with validation that limits `ObjectMeta` to labels and annotations
- custom struct as in https://github.com/kubernetes/kubernetes/pull/91680#discussion_r436014707

pohly replied:
> edit: I was thinking of the PV type, not the PVC type, when I was concerned about complexity. 

That's simple then: PV does not change at all, nor does the code which handles provisioning for the PVC.

> However, the questions about spawning N objects from pod creation still remain.

That's a numbers game. How large is N and how many pods will use several inline volumes? We don't know yet and it may entirely depend on how quick provisioning is in practice, so we have to build something, ask people to try it out and gather feedback.

But even if it turns out to not scale well, it should still be useful for long-lived pods with few replicas (the memcached example).

pohly replied:
I pushed a version with strict validation: only labels and annotations are allowed in the embedded ObjectMeta.

While looking for a way to do this via whitelisting those fields (less error-prone than blacklisting), I found a precedent and used a generalized version of that code: https://github.com/kubernetes/kubernetes/blob/442a69c3bdf6fe8e525b05887e57d89db1e2f3a5/pkg/apis/core/validation/validation.go#L2649-L2666

pohly replied:
Another data point: if we go with embedded labels + annotations, "kustomize" needs a slight update, otherwise it does not apply common labels and annotations to the new fields. It does not matter whether we use ObjectMeta or a custom struct. On the bright side, I looked into this and have a patch ready for kustomize if we decide to add this API.

pohly replied:
We decided to embed `ObjectMeta` + `PersistentVolumeClaimSpec`, so I consider this resolved.

pohly commented:
/retest

pohly commented:
/label api-review

This PR contains the API change from https://github.com/kubernetes/enhancements/issues/1698. As noted in https://github.com/kubernetes/kubernetes/pull/91680#discussion_r433928160, the actual API change is a bit simpler than the one proposed in the KEP.

pohly commented:
/retest

liggitt commented:
load balancing API review to @msau42 as an initial reviewer, and to @thockin as approver

jsafrane commented:
> Should EphemeralVolumeSource embed a PersistentVolumeTemplate (as in the KEP) or is a PersistentVolumeClaimSpec sufficient (as in this PR)? 

IMO, we should follow StatefulSet approach to keep consistency in the API, i.e. use PersistentVolumeTemplate. I don't like copying pod labels to PVC labels - they're different objects and we should not assume pod labels make sense on PVCs.

jsafrane commented:
> Should EphemeralVolumeSource embed a PersistentVolumeTemplate (as in the KEP) or is a PersistentVolumeClaimSpec sufficient (as in this PR)? 

IMO, we should follow StatefulSet approach to keep consistency in the API, i.e. use PersistentVolumeTemplate. I don't like copying pod labels to PVC labels - they're different objects and we should not assume pod labels make sense on PVCs.

pohly commented:
> IMO, we should follow StatefulSet approach to keep consistency in the API, i.e. use PersistentVolumeTemplate.

If we want to be consistent, then we would have to embed `PersistentVolumeClaim`, because that is what `StatefulSet` uses. Another inconsistency will be that `StatefulSet` uses the `PersistentVolumeClaim.Name` whereas ephemeral inline volumes use the `VolumeSource.Name`. Have you considered that?

So overall, I don't buy the argument that we should do this for the sake of consistency - it's not the same situation.

> I don't like copying pod labels to PVC labels - they're different objects and we should not assume pod labels make sense on PVCs.

That is a concern that I share. It could be addressed by limiting the copying to the recommended labels for deployments, i.e. those with `app.kubernetes.io` as prefix (https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels). Those then make sense for the PVCs.

pohly commented:
@jsafrane: suppose we embed an ObjectMeta, what's your expectation regarding the other fields (name, annotations, ownership references)? Should those also be copied? What about the name, if it is set?

msau42 commented:
/assign @jsafrane @saad-ali

pohly commented:
@msau42 responded to the question about scalability and security in https://github.com/kubernetes/enhancements/pull/1830#discussion_r435663795 (the PR which was supposed to be about PRR - sorry for creating some confusion with the two PRs, but we need to move forward on both fronts, i.e. getting the KEP and the code ready):

> Security: @liggitt do you think the proposed PSP additions can address the concern? The PVCs that are created also will continue to go through storageclass quota.

> Scalability: We did implement some scalability tests for volumes according to the scale testing kep. They are available at https://github.com/kubernetes/perf-tests/tree/master/clusterloader2/testing/experimental/storage/pod-startup and currently support built-in ephemeral types like secrets and configmaps, and PVCs. Once we have an implementation, we can potentially run these tests with the CSI mock driver. I'm not sure we'll be able to match the performance of built-in ephemeral types, but we should understand and document the limits.

pohly commented:
> I'm not sure we'll be able to match the performance of built-in ephemeral types, but we should understand and document the limits.

Let me add that matching that performance isn't even a goal. Instead, we want feature parity (storage capacity tracking, normal API for provisioning, etc.). Performance was a design goal for "CSI ephemeral inline volumes", which is why those are still needed.

pohly committed:
GenericEphemeralVolume: 'make update' for API

smarterclayton commented on pkg/apis/core/types.go:
```diff
@@ -1670,6 +1673,29 @@ type CSIVolumeSource struct {
+	// to the PVC after it has been created.
+	//
+	// Required, must not be nil.
+	VolumeClaim *PersistentVolumeClaimSpec
```
A generic ephemeral volume.... is not persistent.  Why "ephemeral" pointing to something called "persistent"?

smarterclayton replied:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1698-generic-ephemeral-volumes#motivation

I really don't understand this motivation in the KEP in terms of helping me understand what this PR is supposed to change (as a reviewer, consumer of kube, and general critic of things that get added).

The motivation sounds like "we want things that are not ephemeral to be shown as ephemeral", but I don't know *why*.  Can you help me unpack what the intent of that KEP is in more controlled way (so we can update the KEP) and I can frame this PR in terms of what the goal is?

pohly replied:
The goal is to support easy deployment of apps that need storage
- that is provided by a CSI driver and
- only as long as a pod runs.

The KEP mentions adding local storage to applications like memcached as use case for this. Here storage serves as an extension or replacement for DRAM. Persistency is not needed and just makes deployment harder.

None of the existing mechanisms fit that goal:
- EmptyDir isn't flexible enough (cannot be provided through a CSI driver).
- Deployment, daemonset, replicas all don't support per-pod volumes. Statefulset is the only one which does, but doesn't delete volumes, i.e. additional logic on top of it is needed to remove them, and it might not have the right semantic for the app (for example, the KubeDB memcached operator uses a deployment).
- CSI ephemeral inline volumes don't support storage capacity tracking and pod startup can fail badly when pods land on a node with insufficient storage; this is by design and would be very hard to fix.

Regarding the naming: if you replace "PersistentVolumeClaim" with "VolumeClaim", the proposal makes sense again. The name is basically historic and just cannot be changed anymore. The old name implies a semantic that just isn't there: the volume always was meant to persist only as long as the corresponding claim (ignore retention policy for the sake of simplicity), and that still holds here. The only difference is that there is a predetermined trigger for deleting the PVC.

Here's an example of deploying memcached with a CSI ephemeral inline volume:
https://github.com/intel/pmem-csi/blob/devel/examples/memcached.md => https://github.com/intel/pmem-csi/blob/devel/deploy/kustomize/memcached/ephemeral/memcached-ephemeral.yaml

CSI ephemeral inline volumes kind of work, but only because in PMEM-CSI a scheduler extender tries to mitigate the risk of bad scheduling decisions. It cannot prevent them entirely and is hard to enable in a cluster. Other CSI  drivers don't have such a Kubernetes-specific extension.

jsafrane replied:
@smarterclayton, maybe it could be emphasized better in the KEP, but the overall motivation distilled into two words is "better emptyDir" - when the root disk is too slow / too small / too shared for pod's ephemeral data.

Yes, using "PersistentVolumeClaim" for something that's called "ephemeral" is not great. We spent quite some time with this in sig-storage and generic ephemeral volume was the best candidate. Suggestions for better naming are welcome.

It has also some other use cases (you can populate it with a restored snapshot / clone a volume instead), still, it's just a scratch space with emptyDir lifetime.

smarterclayton commented:
/hold

i think jordan and I have a lot of questions that we'll need to dig into.

pohly commented:
/retest

pohly commented:
I've rebased and replaced the term "whitelist" with "accept list" in one place.

pohly commented:
/test pull-kubernetes-integration

pohly commented:
/retest

smarterclayton commented on pkg/apis/core/types.go:
```diff
@@ -1670,6 +1673,42 @@ type CSIVolumeSource struct {
+	// inside the template. All other meta data in it is ignored.
+	//
+	// This field is read-only and no changes will be made by Kubernetes
+	// to the PVC after it has been created.
```
I assume scheduling will be blocked until the claimed template is created?

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1746,6 +1749,42 @@ type CSIVolumeSource struct {
+	// The name of the PVC will be `<pod name>-<volume name>` where `<volume name>`
+	// is the name from the `PodSpec.Volumes` array entry. Care must
+	// be taken to keep this concatendated name shorter enough that it
+	// can be used as object name.
```
Pod validation will flag that?

smarterclayton replied:
Controllers will properly summarize that failure?

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1746,6 +1749,42 @@ type CSIVolumeSource struct {
+
+// PersistentVolumeClaimTemplate is used to produce
+// PersistentVolumeClaim objects.
+type PersistentVolumeClaimTemplate struct {
```
Need more description on all of these fields, describing how the object is used.  I should be able to read this godoc and not have to read any other documentation to leverage this.

pohly replied:
So it's better to describe the naming and ownership behavior here instead of under `EphemeralVolumeSource`? Okay, let me move the text.

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1746,6 +1749,42 @@ type CSIVolumeSource struct {
+// PersistentVolumeClaim objects.
+type PersistentVolumeClaimTemplate struct {
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
```
You need to describe which fields are used and ignored.

pohly replied:
That was under `EphemeralVolumeSource`. I agree, it makes more sense here.

smarterclayton replied:
In general it's ok to be specific here, and reference or cover use case driven (duplicate / summarize / highlight specific fields) in the field apidoc.  That way you look at the field to figure out how to use the object as a whole, and can drill down to specific fields if necessary.

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -156,6 +156,9 @@ type VolumeSource struct {
 	// CSI (Container Storage Interface) represents ephemeral storage that is handled by certain external CSI drivers (Beta feature).
 	// +optional
 	CSI *CSIVolumeSource `json:"csi,omitempty" protobuf:"bytes,28,opt,name=csi"`
+	// Ephemeral represents ephemeral storage that is handled by a normal storage driver (Alpha feature).
```
More description!

Probably describe when you would use this and which volumes would be supported.  Likewise describing the connection between this and pvc in the same struct (and improving that description).  `Use X when A, and Y when B`

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1746,6 +1749,42 @@ type CSIVolumeSource struct {
+	// is the name from the `PodSpec.Volumes` array entry. Care must
+	// be taken to keep this concatendated name shorter enough that it
+	// can be used as object name.
+	//
```
Describe what happens when a user creates the PVC up front but pod isn't listed as an owner.  Does pod fail to schedule, fail to admit, fail to start?  Why or why not would someone want adoption to work?  Could a higher level controller like statefulset controller use this?  Does statefulset conrtoller have to verify that no naming conflict exists with the pod template?

If a mutating admission plugin creates an ephemeral volume on the fly at pod creation, what happens?  I.e. ephemeral volumes were slated for admission style PodPreset / magic injection of secrets (give every pod in this namespace a kerberos identity injected as an ephemeral volume), how would that that admission plugin interoperate with the proposed ephemeral creating controller?

pohly reviewed (commented):
I've pushed one additional commit d09168277e53df135c044746c50af5c8dcb42adf with text changes for the external API.

@smarterclayton : perhaps you can check whether that is what you had in mind still today? Then I can make the same change also for the internal type and squash the commits, then re-generate the generated files.

pohly commented:
@smarterclayton : I've squashed the updated documentation and added validation of the generated PVC name. Does that make the API ready for merging, at least as far as you are concerned?

pohly commented:
/retest

smarterclayton commented on pkg/apis/core/types.go:
```diff
@@ -157,6 +157,31 @@ type VolumeSource struct {
 	// CSI (Container Storage Interface) represents ephemeral storage that is handled by certain external CSI drivers (Beta feature).
 	// +optional
 	CSI *CSIVolumeSource
+	// Ephemeral represents a volume that is handled by a normal storage driver (Alpha feature).
```
`cluster storage driver` maybe?  `normal` doesn't sound right.

pohly replied:
Ack. That's better.

smarterclayton commented on pkg/apis/core/types.go:
```diff
@@ -157,6 +157,31 @@ type VolumeSource struct {
 	// +optional
 	CSI *CSIVolumeSource
+	// Ephemeral represents a volume that is handled by a normal storage driver (Alpha feature).
+	// The volume will be created and deleted together with the pod ("ephemeral volume").
```
> The volume's lifecycle is tied to the pod that defines it - it will be created before the pod starts, and deleted before the pod is removed.

pohly replied:
Ack.

pohly replied:
I remembered that the "before the pod is removed" is only true when using foreground deletion, and even then the PV is going to be deleted after the PVC and thus the pod.

My proposal:
- Keep the current behavior and describe it as "The volume's lifecycle is tied to the pod that defines it - it will be created before the pod starts, and deleted when the pod is removed."
- Discuss for beta whether we can make this stricter (tracked here: https://github.com/kubernetes/enhancements/issues/1698#issuecomment-655344680).

smarterclayton commented on pkg/apis/core/types.go:
```diff
@@ -157,6 +157,31 @@ type VolumeSource struct {
+	//    and PersistentVolumeClaim).
+	//
+	// Use PersistentVolumeClaim or one of the vendor-specific
+	// APIs for persistent volumes.
```
> ... for volumes that persist for longer than the lifecycle of an individual pod.

pohly replied:
Ack.

smarterclayton commented on pkg/apis/core/types.go:
```diff
@@ -157,6 +157,31 @@ type VolumeSource struct {
+	// be used that way - see the documentation of the driver for
+	// more information.
+	//
+	// A pod can use both types of ephemeral volumes and
```
> A pod can use both ephemeral and persistent volumes at the same time as long as each volume claim has a unique name.

That actually is another validation thing - is someone allowed to create a PVC mount that references the name that the ephemeral volume would create?  I.e. pod references PVC `foo-bar-baz` and the pod name is `foo-bar` and there is an ephemeral volume `baz` - is that also rejected?

pohly replied:
It is allowed. Whether it works depends on whether the storage driver supports concurrent access by two different pods. This could lead to problems like one pod starting to use a PVC that was meant to be used by another; that's not different from PVC name conflicts between independent apps that get deployed in the same namespace.

Should we prevent it? I don't think so. The promise is that Kubernetes manages these volumes for the pod, but there is no promise that only one pod gets access. The access boundary is the namespace, just as it is for normal PVCs.

To phrase it differently: that the PVC is visible in the cluster is a feature, not a bug. It enables things like taking a snapshot of the volume (for example, for debugging), resizing it, checking its status, etc.

smarterclayton replied:
Sorry, i mean - within the same pod.  What use case would be enabled by allowing a PVC claim in a pod and an ephemeral definiton in a pod to use the same target PVC name?  If we can't think of one, disallow by default.

pohly replied:
Ah, okay. Yes, that can be prevented during validation. Will add.

pohly replied:
Done and pushed. The error looks like this:
```
volumes[1].persistentVolumeClaim.claimName: Invalid value: "test-pod-ephemeral-volume": must not reference a PVC that gets created for an ephemeral volume
```

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1746,6 +1773,54 @@ type CSIVolumeSource struct {
+	// to the PVC after it has been created.
+	//
+	// Required, must not be nil.
+	VolumeClaimTemplate *PersistentVolumeClaimTemplate `json:"volumeClaimTemplate,omitempty" protobuf:"bytes,1,opt,name=volumeClaimTemplate"`
```
If this is required and may not be nil why is it a pointer?

pohly replied:
Because at some point there might be other ways to parameterize an `EphemeralVolumeSource`. For example, the `EmptyDir` volume type could be replaced with:
```
ephemeral:
   emptyDir: ...
```
with a `EmptyDir` pointer field inside `EphemeralVolumeSource`.

Whether that makes sense and really leads to a better API overall remains to be seen. For now the documentation and code are written such that an ephemeral volume always has the `VolumeClaimTemplate`.

pohly replied:
The beta graduation criteria have a line item for this.

smarterclayton replied:
Ok

pohly commented:
/retest

smarterclayton commented on pkg/apis/core/validation/validation.go:
```diff
@@ -666,6 +666,26 @@ func validateVolumeSource(source *core.VolumeSource, fldPath *field.Path, volNam
+				}
+				if podName != "" {
+					pvcName := podName + "-" + volName
+					for _, msg := range ValidatePersistentVolumeName(pvcName, false) {
```
Not PersistentVolumeClaimName?  Or are the two the same?

smarterclayton commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3252,10 +3288,11 @@ func validatePodIPs(pod *core.Pod) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	vols, vErrs := ValidateVolumes(spec.Volumes, fldPath.Child("volumes"))
+	vols, vErrs := ValidateVolumes(spec.Volumes, podMeta, fldPath.Child("volumes"))
```
Hrm.... I don't know that we should be validating volume names in PodSpec in StatefulSets, Daemonsets, or ReplicaSets (or in other workload sets) because we have no idea what pod name a statefulset is going to choose based on the name.

Instead, I think this validation should only be for Pods created directly for now.  Jobs, for instance, have their own name rules.

pohly replied:
`podMeta` will be empty when validating PodSpec in those other scenarios, in which case the name validation is disabled. So that part should be fine.

However, when adding the nil parameter I unintentionally also added it in `validatePodMetadataAndSpec` instead of passing the real podMeta of the pod. Fixed.

That this slipped through testing showed that the new tests were too low-level. Therefore I moved the test cases into the more complete `TestValidatePod`. While doing so, I learned that the generated name is set already as `podMeta.Name`. That simplifies the checking because it can rely on that value.

I couldn't resist modernizing those tests by using `t.Run` and named test cases.

smarterclayton commented on pkg/apis/core/validation/validation.go:
```diff
@@ -354,6 +354,13 @@ func ValidateVolumes(volumes []core.Volume, podMeta *metav1.ObjectMeta, fldPath
+	for _, vol := range volumes {
+		// Remember which PVCs will be created for this pod.
+		if podMeta != nil && vol.VolumeSource.Ephemeral != nil {
+			allCreatedPVCs.Insert(podMeta.Name + "-" + vol.Name)
```
So this doesn't take into account generated name?  Shouldn't this skip this chunk if pod GenerateName is not empty?

pohly replied:
As I learned now, podMeta.Name *is* always set already when validating a pod:

https://github.com/kubernetes/kubernetes/blob/896da2253c77d15fc551843e515039c6e9c86c99/pkg/apis/core/validation/validation.go#L3116-L3120
https://github.com/kubernetes/kubernetes/blob/896da2253c77d15fc551843e515039c6e9c86c99/pkg/apis/core/validation/validation.go#L328-L339
https://github.com/kubernetes/apimachinery/blob/fb3c024b9997b3ee34b110f9c47216b7e7bf3a40/pkg/api/validation/objectmeta.go#L154-L155

I changed the code such that it skips the check if podMeta is nil (non-Pod validation) or podMeta.Name is empty (an error that gets flagged by `ValidateObjectMeta`).

pohly reviewed (commented):
I've pushed the updated validation code.

pohly commented:
/retest

pohly commented:
/retest

pohly commented:
/retest

smarterclayton commented:
/lgtm
/approve

saad-ali commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1746,6 +1773,54 @@ type CSIVolumeSource struct {
+	// validation.
+	//
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
```
I really think the PVC names should be generated or at least partially generated. Consider the case where two pods specify the same exact name by accident, should they end up using the same PVC? What happens when one of those pods terminates?

Not a blocker for alpha, but we should consider this before beta.

pohly replied:
Two different pods will not use the same PVC. The scheduler binder checks that the PVC with the expected name is owned by the pod before reporting that the ephemeral volume is ready, so one pod will start and the other will be blocked until the first pod terminates and the PVC gets deleted.

Let's discuss this for the next iteration of this feature. I've added a line item for it to https://github.com/kubernetes/enhancements/issues/1698#issuecomment-655344680

saad-ali commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1746,6 +1773,54 @@ type CSIVolumeSource struct {
+	// Specifies a read-only configuration for the volume.
+	// Defaults to false (read/write).
+	// +optional
+	ReadOnly bool `json:"readOnly,omitempty" protobuf:"varint,2,opt,name=readOnly"`
```
I'm not sure if another ReadOnly field is necessary here.

Consider removing for beta unless we have a strong justification.

pohly replied:
I'm fine either way. It's here a) for consistency and b) because some volumes may be conceptually read-only. If we can achieve that differently, then we don't really need the flag.

Added to the TODO list for future discussion.

saad-ali reviewed (commented):
LGTM for alpha. Left some considerations for GA

pohly commented:
Will be merged together with the implementation in https://github.com/kubernetes/kubernetes/pull/92784.
